### ISSUE TITLE
Add local cache, `modes` to `aws_codebuild_project` resource

### ISSUE THREAD

**Issue by amilstead** [terraform-providers/terraform-provider-aws#7643]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS CodeBuild just recently published support for "local" caching on code build projects: https://aws.amazon.com/blogs/devops/improve-build-performance-and-save-time-using-local-caching-in-aws-codebuild/ 

The new "mode" attribute is documented here: https://docs.aws.amazon.com/codebuild/latest/APIReference/API_ProjectCache.html#CodeBuild-Type-ProjectCache-modes

Please update the terraform resource to reflect these new features. 

### New or Affected Resource(s)

* aws_codebuild_project

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
resource "aws_codebuild_project" "example" {
  name          = "test-project"
  description   = "test_codebuild_project"
  build_timeout = "5"
  service_role  = "${aws_iam_role.example.arn}"

  artifacts {
    type = "NO_ARTIFACTS"
  }

  cache {
    type = "LOCAL"
    mode = "LOCAL_DOCKER_LAYER_CACHE"
  }
  ...
}
```

**Comment by bflad** [comment]
Requires [AWS Go SDK v1.17.3](https://github.com/aws/aws-sdk-go/releases/tag/v1.17.3)

**Comment by marcoreni** [comment]
Hey @bflad , it seems that AWS SDK is already at 1.18.5, so I'm assuming that we can start working on this. 

I can try to put something together if no-one is working on this.

**Comment by bflad** [comment]
@marcoreni please feel free 😄

**Comment by smiller171** [comment]
Hey @marcoreni Did you get a chance to take a look at this? If not I might take a stab at it

**Comment by imightybigman** [comment]
Any movement on this one?

**Comment by imightybigman** [comment]
https://github.com/terraform-providers/terraform-provider-aws/pull/8215

First time contributing, if there's any obvious issue please let me know before I attempt the acceptance tests.

**Comment by imightybigman** [comment]
Test updated in that PR

**Comment by imightybigman** [comment]
Any update on this ?

**Comment by AlmogCohen** [comment]
That's such an important feature on AWS CodeBuild to make the build way quicker for docker...eager to see it integrated

**Comment by rkreich** [comment]
If anyone is blocked until the #8215 is merged and would like to use local cache but not have Terraform try to revert it, add the following to the CodeBuild project resource:
```
  lifecycle {
    ignore_changes = ["cache"]
  }
```

**Comment by bflad** [comment]
Support for local cache has been merged and will release with version 2.12.0 of the Terraform AWS Provider, likely later today. Thanks to @imightybigman for the enhancement.

**Comment by bflad** [comment]
This has been released in [version 2.12.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#2120-may-24-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
